### PR TITLE
Allow functions to be called directly from element accesses

### LINKED ISSUE #2525: Error when call function from array of functions
Minimal example:

```ts
function foo(): void {}
const arr = [foo];
arr[0]();
```
Error:
```
ERROR TS2757: Type '() => void' has no call signatures.
```

### PR DESCRIPTION
**Pull request by CountBleck**
Fix #2525 

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

File: tests/compiler/function-array.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+const funcs = [funcA, funcB, funcC, funcD, funcE];
+const results = [369, 124, 61, 8, 116];
+
+for (let i = 0; i < results.length; i++) {
+  const expected = results[i];
+  const actual = funcs[i](123);
+
+  assert(expected == actual);
+}
```
Comment (MaxGraey) [inline tests/compiler/function-array.ts:24]: Could you simplify this example to a minimal example? Also, I guess make sense put in this test: https://github.com/AssemblyScript/assemblyscript/blob/main/tests/compiler/std/array-access.ts

Comment (CountBleck) [reply tests/compiler/function-array.ts]: See the current change.

File: src/compiler.ts [hunk 2]
```diff
@@ -6214,6 +6214,37 @@ export class Compiler extends DiagnosticEmitter {
         // fall-through
       }
 
+      case ElementKind.INDEXSIGNATURE: {
```
Comment (dcodeIO) [inline src/compiler.ts:6217]: The fall-through above, from the CLASS case into the INDEXSIGNATURE case, seems suspicious.

Comment (CountBleck) [reply src/compiler.ts:6206]: Nope. Not at all.

Comment (dcodeIO) [reply src/compiler.ts:6206]: Thanks. There is also the resolver's `resolveCallExpression` that needs to be updated with an `INDEXSIGNATURE` case accordingly, so it mirrors the compiler's `compileCallExpression`.

Comment (CountBleck) [reply src/compiler.ts:6206]: Sorry for the delay; I wasn't able to access my laptop for about a month. I notice that `compileCallExpression` has more cases than `resolveCallExpression`, so is such a change necessary?

Comment (dcodeIO) [reply src/compiler.ts]: I've recently looked at the PR again, and found that there is likely potential for general improvements. All the cases that compile the target expression can probably be unified to in fact compile the target expression (call a local, call a global, call a property, call an index signature), then evaluate the type of it, get the signature, and do the indirect call. However, #2548 changes a bunch of code in close vicinity, so I haven't managed yet to make changes to this PR, which I originally intended.

Suggesting to wait for after the mentioned PR has landed, and then see? Would you like to look into these simplifications yourself?

**Pull request closed without merging**